plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -408,6 +413,29 @@ else if (identity.getId() == TRINO_MERGE_PARTITION_DATA) {
                 deletePredicate);
     }
 
+    private TupleDomain<IcebergColumnHandle> getEffectivePredicate(
+            Schema tableSchema,
+            Map<Integer, Optional<String>> partitionKeys,
+            TupleDomain<IcebergColumnHandle> dynamicFilterPredicate,
+            TupleDomain<IcebergColumnHandle> unenforcedPredicate)
+    {
+        if (dynamicFilterPredicate.isAll() || partitionKeys.isEmpty()) {
+            return unenforcedPredicate.intersect(dynamicFilterPredicate);
+        }
+        Set<IcebergColumnHandle> partitionColumns = partitionKeys.keySet().stream()
+                .map(fieldId -> getColumnHandle(tableSchema.findField(fieldId), typeManager))
+                .collect(toImmutableSet());
+        Supplier<Map<ColumnHandle, NullableValue>> partitionValues = memoize(() -> getPartitionValues(partitionColumns, partitionKeys));
+        if (!partitionMatchesPredicate(partitionColumns, partitionValues, dynamicFilterPredicate)) {
+            return TupleDomain.none();
+        }
+        // Filter out eventual partition columns domains from the dynamic filter because they should be irrelevant at data file level
+        dynamicFilterPredicate = dynamicFilterPredicate
+                .filter((columnHandle, domain) -> !partitionKeys.containsKey(columnHandle.getId()));
```

Comment:
There may be some minor efficiency benefit to this but there's no guarantee that this code will stay the same in future. It's not explicitly tested for and it's not something which is necessary for correctness. So this shouldn't be relied on as a "fix" for files having wrong statistics on partitioning columns.

Reply:
Added a test in `TestIcebergNodeLocalDynamicSplitPruning` to specifically go over explicit partition filters in the dynamic filter.

Reply:
That test is good for testing the 2nd commit. This commit is hard to test for, that's not a problem for landing it though. But we should keep in mind that it's only hiding the problem of wrong statistics on partitioning column and we can't guarantee that this code won't change as there is no correctness impact here.